api/propagation/http/propagators.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package http
+
+import (
+	"go.opentelemetry.io/otel/api/core"
+	dctx "go.opentelemetry.io/otel/api/distributedcontext"
+)
+
+// Supplier is an interface that specifies methods to retrieve and store
+// value for a key to an associated carrier.
+// Get method retrieves the value for a given key.
+// Set method stores the value for a given key.
+type Supplier interface {
+	Get(key string) string
+	Set(key, value string)
+}
+
+// SpanContextPropagator is an interface that specifies methods to
+// convert SpanContext to/from byte array.
+type SpanContextPropagator interface {
+	Inject(core.SpanContext, Supplier)
+	Extract(Supplier) core.SpanContext
```

Comment:
I have only written chained propagators for specific stuff and they don't exactly match the behaviour from otep, since chained extractors are called until we get the valid span context, correlation or baggage. I'm not sure what's the use of the generic chainExtractor that calls all the extractors…

And about spancontext in the context - should the users of propagators get the span context from the extractor and pass it themselves to `Tracer.Start` through some of the relationship-establishing option?